05-data-structures-part2.Rmd

```diff
@@ -1,245 +1,132 @@
-than vectors and other data structures we have considered, but they
-provide powerful capabilities.
-
-
-Data frames can be created manually with the `data.frame` function:
+We learned last time that the columns in a data.frame were vectors, so that our data is reliable and consistent throughout the column. As such, if we want to add a new column, we need to start by making a new vector:
 
 ```{r}
-df <- data.frame(id = c('a', 'b', 'c', 'd', 'e', 'f'), x = 1:6, y = c(214:219))
-df
+newCol <- c(2,3,5,12)
 ```
 
-> ## Challenge 1: Data frames {.challenge}
->
-> Try using the `length` function to query
-> your data frame `df`. Does it give the result
-> you expect?
->
-
-Each column in the data frame is simply a list element, which is why
-when you ask for the `length` of the data frame, it tells you the
-number of columns. If you actually want the number of rows, you can
-use the `nrow` function.
-
-We can add columns or rows to a data.frame using `cbind` or `rbind`
-(these are the two-dimensional equivalents of the `c` function):
-
-## Adding columns to a data frame
-
-To add a column we can use `cbind`:
+We can then add this as a column via:
 
 ```{r}
-df <- cbind(df, 6:1)
-df
+cats <- cbind(cats,  newCol)
 ```
 
-Note that R automatically names the column. We may decide to change
-the name by assigning a value using the `names` function:
+Why didn't this work? Of course, R wants to see one element in our new column for every row in the table:
 
 ```{r}
-names(df)[4] <- 'SixToOne'
+cats
+newCol <- c(4,5,8)
+cats <- cbind(cats, newCol)
+cats
 ```
 
-We can also provide a name when we add the column:
+Our new column has appeared, but it's got that ugly name at the top; let's give it something a little easier to understand:
 
 ```{r}
-df <- cbind(df, caps=LETTERS[1:6])
-df
+names(cats)[4] <- 'age'
 ```
 
-(`LETTERS` and `letters` are built-in constants.)
-
-
-## Adding rows to a data frame 
-
-To add a row we use `rbind`:
+Now how about adding rows - in this case, we saw last time that the rows of a data.frame are made of lists:
 
 ```{r}
-df <- rbind(df, list("g", 11, 42, 0, "G"))
+newRow <- list("tortoise", 3.3, TRUE, 9)
+cats <- rbind(cats, newRow)
 ```
 
-Note that we add the row as a list, because we have multiple types
-across the columns. Nevertheless, this doesn't work as expected! What
-do the error messages tell us?
-
-It appears that R was trying to append "g" and "G" as factor
-levels. Why? Let's examine the first column.  We can access a column
-in a `data.frame` by using the `$` operator.
-
+Another thing to look out for has emerged - when R creates a factor, it only allows whatever is originally there when our data was first loaded, which was 'black', 'calico' and 'tabby' in our case. Anything new that doesn't fit into one of its categories is rejected as nonsense, until we explicitly add that as a *level* in the factor:
 
 ```{r}
-class(df$id)
-str(df)
+levels(cats$coat)
+levels(cats$coat) <- c(levels(cats$coat), 'tortoise')
```

Comment:
Pedantic, where I come from tortoise is an animal, tortoiseshell is the colour. Happy to be corrected if that's not an international convention.